server/src/main/java/org/opensearch/gateway/remote/RemoteObject.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote;
+
+import org.opensearch.common.CheckedRunnable;
+import org.opensearch.core.action.ActionListener;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * An interface to read/write and object from/to a remote storage. This interface is agnostic of the remote storage type.
+ * @param <T> The object type which can be upload to or download from remote storage.
+ */
+public interface RemoteObject<T> {
```

Comment:
`RemoteObject` looks more like an entity or POJO. Separate out the entity part 
```
    public String clusterUUID();

    public InputStream serialize() throws IOException;

    public T deserialize(InputStream inputStream) throws IOException;

```

from the actor 

```
    public CheckedRunnable<IOException> writeAsync(ActionListener<Void> listener);

    public T read() throws IOException;

    public void readAsync(ActionListener<T> listener);

    public T get();

```

Reply:
Can `CheckedRunnable<IOException>` be wrapped into the ActionListener itself? or this is more around synchronous error handling purpose?
```
 public CheckedRunnable<IOException> writeAsync(ActionListener<Void> listener);
 
 ```

Reply:
Regarding separating out the methods, I initially created them in different classes like below:
```
public interface RemoteObject <T> {
    public T get();
    public String clusterUUID();
    public InputStream serialize() throws IOException;
    public T deserialize(InputStream inputStream) throws IOException;
}
```

```
public interface RemoteObjectStore <T> {
    public CheckedRunnable<IOException> writeAsync(ActionListener<Void> listener);
    public T read() throws IOException;
    public void readAsync(ActionListener<T> listener);
}
```

With this above design, for uploading IndexMetadata for example I had to write implementations `RemoteIndexMetadata` (implements RemoteObject) and `RemoteIndexMetadataObjectStore` (implements RemoteObjectStore).
The separate RemoteIndexMetadata class make sense as it can differ across every object. 
But for RemoteIndexMetadataObjectStore it seemed that I was creating lot of redundant instances as the internal logic across the instances would be the same for different objects. RemoteObjectStore implementations would only differ in the parameterized entity that they return. So creating new instances of RemoteObjectStore for every object to be uploaded, would make it a little difficult to manage RemoteObjectStore implementations. It would be easier to separate if the RemoteObjectStore is not a parameterized class so we can create a single instance.

How do you suggest to separate out the methods ?

Reply:
`CheckedRunnable<IOException>` can be removed from method signature

Reply:
We might need a singleton instance of `RemoteObjectStore ` that works on different instances of `RemoteObject`
Can we then have
```
public interface RemoteObjectStore <T extends RemoteObjectStore > {
    public CheckedRunnable<IOException> writeAsync(ActionListener<Void> listener);
    public T read() throws IOException;
    public void readAsync(ActionListener<T> listener);
}
```

Reply:
We would have to keep singleton instance of RemoteObjectStore for each concrete implementation of RemoteObject. I can create a factory to vend these instances.

Reply:
I agree on the proposal to keep RemoteObject as the POJO/ entity and the Store related operations (read/write) via a separate interface and that can be singleton instance. In future lets say when we have different RemoteStore implementation, the entity may remain same.